### PR TITLE
Note sorting algorithm rework

### PR DESCRIPTION
**Pull request by zonkmachine**
Addresses some of the issues in https://github.com/LMMS/lmms/issues/3880

 Sorting after position and key. Probably only needed with the arpeggiator. This works well and doesn't seem to use a whole lot cpu cycles. The notes are sorted on input from mouse or keyboard in ` Pattern::addNote()`. Actions on the notes in the Piano Roll also triggers a sort via `Pattern::rearrangeAllNotes()` with a helper function in note.h . The sorting is duplicated when adding notes in the Piano Roll and this has its own post in https://github.com/LMMS/lmms/issues/3880 but is probably a wontfix.

I've pulled these against stable-1.2 but they can just as well go into master or be split over the two.

### CODE REVIEW

**Comment by zonkmachine** [comment]
The simplest approach here may be to just duplicate the position/key sorting algorithm for Pattern::addNote().

**Comment by zonkmachine** [comment]
The latest (third) commit reverts the first one and implement a position/key note sort for `Pattern::addNote()`

Here is an example output from playing chords and single notes with a keyboard.
```
        <pattern steps="16" muted="0" type="1" name="TripleOscillator" pos="0">
          <note pan="0" key="41" vol="173" pos="72" len="96"/>
          <note pan="0" key="47" vol="173" pos="72" len="96"/>
          <note pan="0" key="50" vol="168" pos="72" len="96"/>
          <note pan="0" key="50" vol="179" pos="252" len="84"/>
          <note pan="0" key="45" vol="173" pos="264" len="84"/>
          <note pan="0" key="53" vol="165" pos="264" len="84"/>
          <note pan="0" key="41" vol="182" pos="420" len="84"/>
          <note pan="0" key="47" vol="176" pos="468" len="24"/>
          <note pan="0" key="50" vol="168" pos="540" len="24"/>
        </pattern>
```

**Comment by zonkmachine** [comment]
Note that old patches will open up like before and all will be fine. The next time you edit a track it will be sorted with the new method and this will probably be only an advantage. Probably. Maybe you like the notes to shift around a bit, and you'll only really notice this in arpeggiated tracks with more than two simultaneous notes.
If this is not wanted you can just _open/press a note in the piano roll to trigger a sort/save_ the project in 1.1.3 and it'll go wonky again.

**Comment by zonkmachine** [comment]
Got a random crash. I don't think it's related to this PR though. The project had been running for maybe an hour or so. I was looping two 3osc tracks with an automation track for the Master pitch, two notes detuned in the one track and an LFO detuning the other synth.
`m_frequencyNeedsUpdate = true;` LFO ... or Note detune ... or Master pitch?

[bt](https://gist.github.com/zonkmachine/211524d32e670682c8afbdaa1d2c6ccc#file-gistfile1-txt-L4) / [bt full](https://gist.github.com/zonkmachine/211524d32e670682c8afbdaa1d2c6ccc#file-gistfile1-txt-L51)
```
Program terminated with signal SIGSEGV, Segmentation fault.
#0  0x000000000062e0ea in NotePlayHandle::setFrequencyUpdate (this=0x81) at /home/zonkmachine/builds/lmms/lmms/include/NotePlayHandle.h:263
263			m_frequencyNeedsUpdate = true;
(gdb) bt
#0  0x000000000062e0ea in NotePlayHandle::setFrequencyUpdate (this=0x81) at /home/zonkmachine/builds/lmms/lmms/include/NotePlayHandle.h:263
#1  0x00000000006270a1 in InstrumentTrack::updateBaseNote (this=0x7fce29e77620) at /home/zonkmachine/builds/lmms/lmms/src/tracks/InstrumentTrack.cpp:543
#2  0x000000000062710a in InstrumentTrack::updatePitch (this=0x7fce29e77620) at /home/zonkmachine/builds/lmms/lmms/src/tracks/InstrumentTrack.cpp:552
#3  0x000000000063f89f in InstrumentTrack::qt_static_metacall (_o=0x7fce29e77620, _c=QMetaObject::InvokeMetaMethod, _id=6, _a=0x7fffe6625550)
    at /home/zonkmachine/builds/lmms/lmms/build/src/moc_InstrumentTrack.cpp:68
#4  0x00007fce35cf187a in QMetaObject::activate (sender=0x7fce29e78898, m=<optimized out>, local_signal_index=<optimized out>, argv=0x0)
    at kernel/qobject.cpp:3539
#5  0x0000000000642f4b in Model::dataChanged (this=0x7fce29e78898) at /home/zonkmachine/builds/lmms/lmms/build/src/moc_Model.cpp:102
#6  0x0000000000642e09 in Model::qt_static_metacall (_o=0x7fce29e78898, _c=QMetaObject::InvokeMetaMethod, _id=0, _a=0x7fffe66256b0)
    at /home/zonkmachine/builds/lmms/lmms/build/src/moc_Model.cpp:51
#7  0x00007fce35cf187a in QMetaObject::activate (sender=0x2ee0210, m=<optimized out>, local_signal_index=<optimized out>, argv=0x0)
    at kernel/qobject.cpp:3539
#8  0x000000000063b4ab in ControllerConnection::valueChanged (this=0x2ee0210)
    at /home/zonkmachine/builds/lmms/lmms/build/src/moc_ControllerConnection.cpp:104
#9  0x000000000063b354 in ControllerConnection::qt_static_metacall (_o=0x2ee0210, _c=QMetaObject::InvokeMetaMethod, _id=0, _a=0x7fcde40521b0)
    at /home/zonkmachine/builds/lmms/lmms/build/src/moc_ControllerConnection.cpp:52
```

**Comment by PhysSong** [comment]
`NotePlayHandle::setFrequencyUpdate (this=0x81)` was called in
https://github.com/LMMS/lmms/blob/54f3eccad77103ff7441470cbfc5739180f3c1ba/src/tracks/InstrumentTrack.cpp#L543
and it may indicate an invalid reference of iterator occured.

**Comment by PhysSong** [comment]
`m_processHandles` is used in non-thread-safe way(i.e. without proper synchronization). `(this=0x81)` means `*it == 0x81`, which is an invalid address.
Anyway, I'm pretty sure it's unrelated with this PR. For this kind of crashes, I'll open a separate issue. :)

**Comment by gi0e5b06** [comment]
#3879 seems related to this NotePlayHandle race condition.

**Comment by zonkmachine** [comment]
This should go into master. I'll switch it over.

**Comment by zonkmachine** [comment]
I switched back to stable 1.2 and this is ready for a review. Switching sorting order was a matter of flipping over two 'lesser than' signs and this is now less buggy than before. See the note on backward compatibility [here](https://github.com/LMMS/lmms/pull/3881#issuecomment-336606142).

File: src/tracks/Pattern.cpp [hunk 1]
```diff
@@ -213,7 +213,7 @@ Note * Pattern::addNote( const Note & _new_note, const bool _quant_pos )
 	}
 
 	instrumentTrack()->lock();
-	if( m_notes.size() == 0 || m_notes.back()->pos() <= new_note->pos() )
+	if( m_notes.size() == 0 || m_notes.back()->pos() < new_note->pos() )
```
Comment (lukas-w) [inline src/tracks/Pattern.cpp]: As far as I understand the code, all lines 216 to 245 do is insert the new note at the right position (sorted by the criteria in `lessThan`). If so, we can reduce this to a single line:
```cpp
m_notes.insert(std::upper_bound(m_notes.begin(), m_notes.end(), new_note, Note:lessThan), new_note);
```
This reduces code complexity, improves performance by doing a binary search, and removes the duplicate compare code by using the `lessThan` method.

Comment (zonkmachine) [reply src/tracks/Pattern.cpp]: > As far as I understand the code, all lines 216 to 245 do is insert the new note at the right position (sorted by the criteria in lessThan).

Yes indeed. Oneliner, nice but I can't make it work.

Computer says no. 8)

```cpp
/usr/include/c++/4.8/bits/stl_algo.h: In instantiation of ‘_FIter std::upper_bound(_FIter, _FIter, const _Tp&, _Compare) [with _FIter = Note**; _Tp = Note*; _Compare = bool (*)(Note*&, Note*&)]’:
/home/zonkmachine/builds/lmms/lmms/src/tracks/Pattern.cpp:216:90:   required from here
/usr/include/c++/4.8/bits/stl_algo.h:2543:31: error: invalid initialization of reference of type ‘Note*&’ from expression of type ‘Note* const’
    if (__comp(__val, *__middle))
                               ^
```

Comment (lukas-w) [reply src/tracks/Pattern.cpp]: Ah, you'll have to change `Note::lessThan` to take `const` pointers.

Comment (zonkmachine) [reply src/tracks/Pattern.cpp]: Perfect, thanks!

**Comment by zonkmachine** [comment]
I took the opportunity to practice and updated the other sorting function to std::sort() as qSort() is obsolete.

Now I think I'm done. The notes are correctly sorted for both functions and there is no code changes needed to save/load the patterns correctly.

**Pull request merged**